### PR TITLE
issue #1 Create Splash Activity

### PR DESCRIPTION
**Pull request by jamacado**
**Create Splash Activity:** #1 

- Se agrega coroutines-android al depencies, se crea interface Scope para manejar las coroutines.
- Se crea el package splash y se implementa la activity y el viewmodel.

Screenshot_3

**Testing Done:**

- On Device: Nexus 6P

How To Test:

Signed-off-by:

Javier M. Cardona <javier.mcardona@gmail.com>

### CODE REVIEW

File: presentation/src/main/java/com/architect/coders/mu8/splash/SplashActivity.kt [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_splash)
+        tvVersion = findViewById(R.id.tv_version)
+        viewModel = ViewModelProviders.of(this)[SplashViewModel::class.java]
+
+        viewModel.model.observe(this, Observer(::initUi))
+    }
+
+    fun initUi(model: UiModel) {
```
Comment (SalosV) [inline presentation/src/main/java/com/architect/coders/mu8/splash/SplashActivity.kt]: Hacer esta función privada

File: presentation/src/main/java/com/architect/coders/mu8/splash/SplashViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+import kotlinx.coroutines.launch
+
+class SplashViewModel : ViewModel(), Scope by Scope.Impl() {
+
+    companion object {
+        const val TIME_SLEEP: Long = 5000
+    }
+    sealed class UiModel {
+        class GetVersion(val versionName: String) : UiModel()
+        class Navigation : UiModel()
```
Comment (SalosV) [inline presentation/src/main/java/com/architect/coders/mu8/splash/SplashViewModel.kt]: Como no tiene estado Navigation puede ser un object

File: presentation/src/main/java/com/architect/coders/mu8/splash/SplashActivity.kt [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+class SplashActivity : AppCompatActivity() {
+
+    private lateinit var viewModel: SplashViewModel
+    private lateinit var tvVersion: TextView
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_splash)
+        tvVersion = findViewById(R.id.tv_version)
```
Comment (recaldev) [inline presentation/src/main/java/com/architect/coders/mu8/splash/SplashActivity.kt:22]: Add a new line above this to improve reading.

**Review by recaldev — CHANGES_REQUESTED** [review]
> **Create Splash Activity:** #1
> 
> * Se agrega coroutines-android al depencies, se crea interface Scope para manejar las coroutines.
> * Se crea el package splash y se implementa la activity y el viewmodel.
> 
> Screenshot_3
> 
> **Testing Done:**
> 
> * On Device: Nexus 6P
> 
>** How To Test:**
>
>Run app, the splash activity is launcher the app
> 
> Signed-off-by:
> 
> Javier M. Cardona [javier.mcardona@gmail.com](mailto:javier.mcardona@gmail.com)

@jamacado I don't see anything written in the how to test section, can you complete it?

File: presentation/src/main/java/com/architect/coders/mu8/splash/SplashViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import com.architect.coders.mu8.BuildConfig
+import com.architect.coders.mu8.utils.Scope
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+
+class SplashViewModel : ViewModel(), Scope by Scope.Impl() {
+
+    companion object {
```
Comment (recaldev) [inline presentation/src/main/java/com/architect/coders/mu8/splash/SplashViewModel.kt]: Companion object should go at the end of the class.

File: presentation/src/main/java/com/architect/coders/mu8/splash/SplashViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        const val TIME_SLEEP: Long = 5000
+    }
+    sealed class UiModel {
```
Comment (recaldev) [inline presentation/src/main/java/com/architect/coders/mu8/splash/SplashViewModel.kt:15]: Remember to separate methods/Inner Classes with one line.

File: presentation/src/main/java/com/architect/coders/mu8/splash/SplashViewModel.kt [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+    init {
+        initScope()
+    }
+
+    private fun onGetVersion() {
+        _model.value = UiModel.GetVersion(BuildConfig.VERSION_NAME)
+        launch {
+            delay(TIME_SLEEP)
+            onNext()
```
Comment (recaldev) [inline presentation/src/main/java/com/architect/coders/mu8/splash/SplashViewModel.kt]: Rename this method, think of a name that explains what do you want to do next.

File: presentation/src/main/java/com/architect/coders/mu8/splash/SplashViewModel.kt [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+            if (_model.value == null) onGetVersion()
+            return _model
+        }
+
+    init {
+        initScope()
+    }
+
+    private fun onGetVersion() {
+        _model.value = UiModel.GetVersion(BuildConfig.VERSION_NAME)
```
Comment (recaldev) [inline presentation/src/main/java/com/architect/coders/mu8/splash/SplashViewModel.kt]: You can put this inside the launch.

File: presentation/src/main/java/com/architect/coders/mu8/splash/SplashViewModel.kt [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+        launch {
+            delay(TIME_SLEEP)
+            onNext()
+        }
+    }
+
+    fun onNext() {
```
Comment (recaldev) [inline presentation/src/main/java/com/architect/coders/mu8/splash/SplashViewModel.kt]: This function can be private.

File: presentation/src/main/java/com/architect/coders/mu8/utils/PresentationExtensions.kt [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+import android.view.ViewGroup
+import android.widget.ImageView
+import androidx.annotation.LayoutRes
+import androidx.fragment.app.FragmentActivity
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.ViewModelProvider
+import androidx.lifecycle.ViewModelProviders
+import com.bumptech.glide.Glide
+
+    fun ImageView.loadUrl(url: String, placeHolder: Int, errorImage: Int) {
```
Comment (recaldev) [inline presentation/src/main/java/com/architect/coders/mu8/utils/PresentationExtensions.kt]: Why this class have added spaces at the beggining?

File: presentation/src/main/res/values/colors.xml [hunk 10]
```diff
@@ -3,4 +3,6 @@
     <color name="colorPrimary">#008577</color>
     <color name="colorPrimaryDark">#00574B</color>
     <color name="colorAccent">#D81B60</color>
+
+    <color name="redMarvel">#ED1C24</color>
```
Comment (recaldev) [inline presentation/src/main/res/values/colors.xml]: You should be using the primary color, check @SalosV 's PR, I think we defined those colors already.

File: presentation/src/main/res/layout/activity_splash.xml [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+    android:background="@color/redMarvel"
+    tools:context=".splash.SplashActivity">
+
+    <TextView
+        android:id="@+id/tv_version"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginStart="14dp"
+        android:layout_marginEnd="14dp"
+        android:layout_marginBottom="20dp"
```
Comment (recaldev) [inline presentation/src/main/res/layout/activity_splash.xml]: Create dimens for this values

File: presentation/src/main/java/com/architect/coders/mu8/splash/SplashViewModel.kt (hunk 7 above)
Comment (recaldev) [inline presentation/src/main/java/com/architect/coders/mu8/splash/SplashViewModel.kt]: @jamacado the version is missing some info, check the initial sketch we had: 

MU8

File: presentation/src/main/res/layout/activity_splash.xml [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+
+    <ImageView
+        android:id="@+id/img_logo"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        app:layout_constraintBottom_toBottomOf="parent"
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toTopOf="parent"
+        app:srcCompat="@drawable/marvel_comic" />
```
Comment (recaldev) [inline presentation/src/main/res/layout/activity_splash.xml]: Why use the compat version if you can use `src` directly?

**Review by recaldev — CHANGES_REQUESTED** [review]

File: presentation/src/main/java/com/architect/coders/mu8/utils/PresentationExtensions.kt (hunk 9 above)
Comment (jamacado) [reply presentation/src/main/java/com/architect/coders/mu8/utils/PresentationExtensions.kt]: detekt automatic change the class

File: presentation/src/main/res/values/colors.xml (hunk 10 above)
Comment (jamacado) [reply presentation/src/main/res/values/colors.xml]: el que utiliza @SalosV me parece mas apagado, creo que se ve mejor con uno mas vivo

File: presentation/src/main/java/com/architect/coders/mu8/splash/SplashActivity.kt (hunk 3 above)
Comment (jamacado) [reply presentation/src/main/java/com/architect/coders/mu8/splash/SplashActivity.kt:22]: ok, new line

File: presentation/src/main/java/com/architect/coders/mu8/splash/SplashViewModel.kt (hunk 4 above)
Comment (jamacado) [reply presentation/src/main/java/com/architect/coders/mu8/splash/SplashViewModel.kt]: move companion object end of the class

File: presentation/src/main/java/com/architect/coders/mu8/splash/SplashViewModel.kt (hunk 6 above)
Comment (jamacado) [reply presentation/src/main/java/com/architect/coders/mu8/splash/SplashViewModel.kt]: change name onChangeOfNavigationStatus

File: presentation/src/main/java/com/architect/coders/mu8/splash/SplashViewModel.kt (hunk 7 above)
Comment (jamacado) [reply presentation/src/main/java/com/architect/coders/mu8/splash/SplashViewModel.kt]: ok

File: presentation/src/main/res/layout/activity_splash.xml (hunk 12 above)
Comment (jamacado) [reply presentation/src/main/res/layout/activity_splash.xml]: ok

File: presentation/src/main/java/com/architect/coders/mu8/splash/SplashViewModel.kt (hunk 5 above)
Comment (jamacado) [reply presentation/src/main/java/com/architect/coders/mu8/splash/SplashViewModel.kt:15]: ok

File: presentation/src/main/java/com/architect/coders/mu8/splash/SplashActivity.kt (hunk 3 above)
Comment (recaldev) [reply presentation/src/main/java/com/architect/coders/mu8/splash/SplashActivity.kt:22]: @jamacado the line was after the `setContentView`

File: presentation/src/main/java/com/architect/coders/mu8/utils/PresentationExtensions.kt (hunk 9 above)
Comment (recaldev) [reply presentation/src/main/java/com/architect/coders/mu8/utils/PresentationExtensions.kt]: That's not possible, there is no rule or style guide that suggest to do this, please fix this.

File: presentation/src/main/res/values/colors.xml (hunk 10 above)
Comment (recaldev) [reply presentation/src/main/res/values/colors.xml]: We can have this discussion later, for now let's sync up and use the same color.

**Review by recaldev — CHANGES_REQUESTED** [review]

File: presentation/src/main/java/com/architect/coders/mu8/splash/SplashActivity.kt [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+
+        setContentView(R.layout.activity_splash)
+        tvVersion = findViewById(R.id.tv_version)
+        viewModel = ViewModelProviders.of(this)[SplashViewModel::class.java]
+
+        viewModel.model.observe(this, Observer(::initUi))
+    }
+
+    private fun initUi(model: UiModel) {
+
```
Comment (recaldev) [inline presentation/src/main/java/com/architect/coders/mu8/splash/SplashActivity.kt]: Remove extra space.

File: presentation/src/main/res/values/dimens.xml [hunk 14]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+
+    <dimen name="margin14">14dp</dimen>
+    <dimen name="maargin20">20dp</dimen>
```
Comment (recaldev) [inline presentation/src/main/res/values/dimens.xml]: Misspelled `maargin`, also, please add a underscore before the number.

**Review by recaldev — CHANGES_REQUESTED** [review]

**Comment by recaldev** [comment]
@jamacado I found a bug on your implementation, if you wait until the splash send you to the home screen and you hit the back button you get the splash again and this is wrong, you have to call finish on your Activity when launching a new one.

File: presentation/src/main/java/com/architect/coders/mu8/splash/SplashViewModel.kt [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+    private fun onStringVersion(): String {
+        return BuildConfig.VERSION_NAME
+    }
+
+    override fun onCleared() {
+        destroyScope()
+        super.onCleared()
+    }
+
+    companion object {
```
Comment (aliceresponde) [inline presentation/src/main/java/com/architect/coders/mu8/splash/SplashViewModel.kt]: remove companion, is not needed.  use a constant

File: presentation/src/main/res/values/dimens.xml (hunk 14 above)
Comment (aliceresponde) [reply presentation/src/main/res/values/dimens.xml]: I suggest use  this convention:
```
<dimen name="dp_14">14dp</dimen>
<dimen name="dp_20">20dp</dimen>
```

File: presentation/src/main/java/com/architect/coders/mu8/utils/PresentationExtensions.kt (hunk 9 above)
Comment (aliceresponde) [reply presentation/src/main/java/com/architect/coders/mu8/utils/PresentationExtensions.kt]: format the code

File: presentation/src/main/java/com/architect/coders/mu8/utils/PresentationExtensions.kt [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+
+    inline fun <reified T : ViewModel> FragmentActivity.getViewModel(crossinline factory: () -> T): T {
+        val vmFactory = object : ViewModelProvider.Factory {
+            override fun <U : ViewModel> create(modelClass: Class<U>): U = factory() as U
+        }
+        return ViewModelProviders.of(this, vmFactory)[T::class.java]
+    }
+
+    fun String.getVersion(): String {
+        return "V$this"
```
Comment (aliceresponde) [inline presentation/src/main/java/com/architect/coders/mu8/utils/PresentationExtensions.kt]: I think its not a valid extension fun. You will not use it.

**Comment by jamacado** [comment]
> 
> 
> @jamacado I found a bug on your implementation, if you wait until the splash send you to the home screen and you hit the back button you get the splash again and this is wrong, you have to call finish on your Activity when launching a new one.
@recaldev 
ok,  finish is added
Question
is better finish o flag into intent?

**Review by recaldev — APPROVED** [review]

**Review by aliceresponde — APPROVED** [review]
lg

**Pull request merged**